Make all identifiers dynamically-sized types (wrappers around `str`)

jplatte opened the pull request:
Resolves #417.

jplatte committed:
identifiers: Make DeviceKeyId a DST

agraven commented:
Just an absentminded thought, but would it be worth it to have `<Indentifier>Buf` types as either type aliases or wrappers to boxed types to make them a bit easier to work with? Something analogous to `Path` and `PathBuf` in the standard library.

jplatte commented:
We have those on `main`, they have the suffix `Box` instead of `Buf` because they're not buffers (they don't have separate length / capacity like `PathBuf`, `String` and so on) and with Conduit likely wanting to use `Arc` in a bunch of places instead of `Box` and sbd. from Fractal (IIRC) previously stating that they don't really see the point, I was actually looking to remove them. This PR does just that in the first commit (which also has the effect of getting rid of the `paste` dependency).

jplatte committed:
identifiers: Make EventId a DST

jplatte committed:
identifiers: Move opaque identifier declarations out of macro

jplatte committed:
identifiers: Make MxcUri a DST

jplatte committed:
identifiers: Make RoomAliasId a DST

jplatte committed:
identifiers: Make RoomId a DST

jplatte committed:
identifiers: Make RoomIdOrAliasId a DST

jplatte committed:
identifiers: Make UserId a DST

jplatte committed:
identifiers: Make KeyId a DST

jplatte committed:
identifiers: Derive PartialEq, Eq, PartialOrd, Ord, Hash for str newtypes

… instead of generating them in our own macro.

jplatte committed:
identifiers: Clean up internal macros

jplatte committed:
identifiers: Make server_name! macro return a reference instead of an owned value

It is easy to turn it into an owned value through `.to_owned()`.
The reverse was not possible without tying the reference to the lifetime
of a temporary variable.

jplatte committed:
identifiers: Update changelog

jplatte committed:
identifiers: Shorten RoomVersionId variant names

jplatte committed:
identifiers: Rename RoomIdOrAliasId to RoomOrAliasId
